### DISCUSSION TITLE
Exhaustive testing can cause a lot of 'duplicated' testing

### DISCUSSION THREAD

**Discussion by sareninden** [pointfreeco/swift-composable-architecture#479] [Ideas]
Hi,

I am working for a while now with TCA and I noticed that when one action leads to another, and to another and so on might cause a lot of duplication since any path might diverge into any other path. Of course you stub everything to direct it into a certain flow but still I end up with over ten receive actions on my teststore from time to time.

For this we could introduce an option like `terminate` on the TestStore. 

```swift
    public func terminate(
        file: StaticString = #file,
        line: UInt = #line
    ) {
        guard receivedActions.isEmpty == false || longLivingEffects.isEmpty == false else {
            XCTFail(
              """
              Terminate called while no follow up effects are scheduled
              """,
              file: file, line: line
            )
            return
        }
        
        receivedActions.removeAll()
        longLivingEffects.removeAll()
    }
```

This is not ideal since new actions might be inserted before this terminate. For this we could do something like
```swift
    public func receiveAndTerminate(
        _ expectedAction: Action,
        file: StaticString = #file,
        line: UInt = #line,
        _ update: @escaping (inout LocalState) throws -> Void = { _ in }
      ) {
          receive(expectedAction, file: file, line: line, update)
          terminate(file: file, line: line)
      }
```
This would ensure that we terminate with the expected receive command.

The main advantage of this is to reduce test duplication.

For example I now test pressing the send button
actions
* sendFormPressed
* validateForm
* sendForm/showError
* sendFormSucceeded/sendFormFailed
* dismiss

This is even a simplified chain of what I am having. But it also duplicates for example another test where I validate the sendForm (or I would only write a single test that handles the failing and 'assume' that it is clear that another 'unrelated' test test the other scenario).

Alternative
It is also possible to have the environment return an `Effect.none` for some cases. This will also stop the chain. In my opinion this is less clear in its intend and does not improve test readability/intention.

**Comment by stephencelis** [comment]
Thanks for the idea! As heavy users of TCA, we've encountered our fair share of test verboseness :sweat_smile:

In general we think that exhaustivity is a good thing, and that tests that are verbose with a lot of duplicated testing could indicate a complex reducer that could be broken down into smaller reducers that are easier to test.

That said, I could still see the library leveraging something like `XCTSkip*` to allow users to skip the final assertion.

```swift
store.skipCompletion()
```

That way these tests are still called out in some way.

**Reply by stephencelis** [reply]
This came up in another conversation on this topic, but I remembered that if you want to test things non-exhaustively you can always drive a test with a normal `Store` and `ViewStore`, which would allow you to ignore in-flight effects. The downside is you lose the ability to easily asset against received actions, though you can still assert against how they mutate state. Some of this could be recovered with test helpers, but at that point you're starting to rewrite `TestStore` 😅

A `skipCompletion` helper may still be a handy middle ground (though I think we'd still want to add the `XCTSkip*` machinery to call out uses), so we'll look into it.

**Reply by sareninden** [reply]
Thanks for the replies, I will see what you guys will come up with =)
